zerver/lib/user_groups.py

```diff
@@ -88,66 +86,67 @@ def has_user_group_access(
     if not allow_deactivated and user_group.deactivated:
         raise JsonableError(_("User group is deactivated."))
 
-    if as_subgroup:
-        # At this time, we only check for realm ID of a potential subgroup.
-        return True
+    return True
 
-    if for_read and not user_profile.is_guest:
-        # Everyone is allowed to read a user group and check who
-        # are its members. Guests should be unable to reach this
-        # code path, since they can't access user groups API
-        # endpoints, but we check for guests here for defense in
-        # depth.
-        return True
 
-    if user_group.is_system_group:
-        return False
+def get_user_group_by_id_in_realm(
+    user_group_id: int, realm: Realm, *, for_read: bool, allow_deactivated: bool = False
+) -> NamedUserGroup:
+    try:
+        if for_read:
+            user_group = NamedUserGroup.objects.get(id=user_group_id, realm=realm)
+        else:
+            user_group = NamedUserGroup.objects.select_for_update().get(
+                id=user_group_id, realm=realm
+            )
 
-    can_edit_all_user_groups = user_profile.can_edit_all_user_groups()
+        if not allow_deactivated and user_group.deactivated:
+            raise JsonableError(_("User group is deactivated."))
+        return user_group
+    except NamedUserGroup.DoesNotExist:
+        raise JsonableError(_("Invalid user group"))
 
-    group_member_ids = get_user_group_direct_member_ids(user_group)
-    if (
-        not user_profile.is_realm_admin
-        and not user_profile.is_moderator
-        and user_profile.id not in group_member_ids
-    ):
-        can_edit_all_user_groups = False
 
+def check_permission_for_managing_all_groups(
+    user_group: UserGroup, user_profile: UserProfile
+) -> bool:
+    can_edit_all_user_groups = user_profile.can_edit_all_user_groups()
     if can_edit_all_user_groups:
-        return True
+        if user_profile.is_realm_admin or user_profile.is_moderator:
+            return True
 
-    return user_has_permission_for_group_setting(
-        user_group.can_manage_group,
-        user_profile,
-        NamedUserGroup.GROUP_PERMISSION_SETTINGS["can_manage_group"],
-    )
+        group_members = get_user_group_direct_member_ids(user_group)
+        if user_profile.id in group_members:
+            return True
```

Comment:
This requires fetching all 20k users if it's a big group; we should just check if the user is in the group with the function for that.

Reply:
Yes, we should do that. Should we also change this to consider members of recursive subgroups as well instead of just allowing direct members.

Reply:
Yes, of course.